### ISSUE TITLE
Large UDT causes ORA-06502: PL/SQL: numeric or value error: raw variable length too long

### ISSUE THREAD

**Issue by belaw** [oracle/dotnet-db-samples#227]
### Bug Report
- NuGet Package: Oracle.ManagedDataAccess 21.4.0
- Database version: Oracle Database 19c Standard Edition 2 Release 19.0.0.0.0 - Production

---

Hello, we have a stored procedure in our database which has a UDT output parameter.
When we try to call that procedure, we get the following exception:
```
Oracle.ManagedDataAccess.Client.OracleException
ORA-06502: PL/SQL: numeric or value error: raw variable length too long
ORA-06512: at "SYS.DBMS_PICKLER", line 404
ORA-06512: at line 14
   at OracleInternal.UDT.Types.UDTNamedType.GetNamedTypeMetaData(OracleConnection conn, OracleCommand getTypeCmd)
   at OracleInternal.UDT.Types.UDTTypeCache.GetMetaData(OracleConnection conn, UDTNamedType udtType)
   at OracleInternal.UDT.Types.UDTTypeCache.CreateUDTType(OracleConnection conn, String schemaName, String typeName)
   at OracleInternal.UDT.Types.UDTTypeCache.GetUDTType(OracleConnection conn, String schemaName, String typeName)
   at OracleInternal.ConnectionPool.OraclePoolManager.GetUDTType(OracleConnection conn, String schemaName, String typeName)
   at Oracle.ManagedDataAccess.Client.OracleConnection.GetUDTTypeFromCache(String schemaName, String typeName)
   at Oracle.ManagedDataAccess.Client.OracleParameter.GetUDTType(OracleConnection conn)
   at Oracle.ManagedDataAccess.Client.OracleParameter.PreBind_UDT(OracleConnection conn)
   at Oracle.ManagedDataAccess.Client.OracleParameter.PreBind(OracleConnectionImpl connImpl, ColumnDescribeInfo cachedParamMetadata, Boolean& bMetadataModified, Int32 arrayBindCount, ColumnDescribeInfo& paramMetaData, Object& paramValue, Boolean isEFSelectStatement, SqlStatementType stmtType)
   at OracleInternal.ServiceObjects.OracleCommandImpl.InitializeParamInfo(ICollection paramColl, OracleConnectionImpl connectionImpl, ColumnDescribeInfo[] cachedParamMetadata, Boolean& bMetadataModified, Boolean isEFSelectStatement, MarshalBindParameterValueHelper& marshalBindValuesHelper)
   at OracleInternal.ServiceObjects.OracleCommandImpl.ProcessParameters(OracleParameterCollection paramColl, OracleConnectionImpl connectionImpl, ColumnDescribeInfo[] cachedParamMetadata, Boolean& bBindMetadataModified, Boolean isEFSelectStatement, MarshalBindParameterValueHelper& marshalBindValuesHelper)
   at OracleInternal.ServiceObjects.OracleCommandImpl.ExecuteNonQuery(String commandText, OracleParameterCollection paramColl, CommandType commandType, OracleConnectionImpl connectionImpl, Int32 longFetchSize, Int64 clientInitialLOBFS, OracleDependencyImpl orclDependencyImpl, Int64[]& scnFromExecution, OracleParameterCollection& bindByPositionParamColl, Boolean& bBindParamPresent, OracleException& exceptionForArrayBindDML, OracleConnection connection, Boolean isFromEF)
   at Oracle.ManagedDataAccess.Client.OracleCommand.ExecuteNonQuery()
...
```

The method `CreateGetUDTTypeCmd` in your NuGet package creates an `OracleCommand` that calls the `sys.dbms_pickler.get_type_shape` stored procedure. **The method sets the size of the parameter `TDS` to 4000. I believe that that is where the bug is.**
I looked at the `sys.dbms_pickler.get_type_shape` code in our database:

```sql
FUNCTION GET_TYPE_SHAPE (
...
        TDS OUT LONG RAW, -- The size of this parameter is set to 4000
...
) 
  RETURN BINARY_INTEGER IS
  OIDBUF RAW(4000);
  TDSBUF RAW(32000); -- This is a buffer with the size of 32000
  RC BINARY_INTEGER;
BEGIN
  RC := GET_DESCRIPTORS(FULLTYPENAME, OIDBUF, VERSION, TDSBUF); -- The buffer is filled with data
...
    TDS := TDSBUF; -- This is line 404 - The buffer of size 32000 is being written to the parameter of size 4000
...
  RETURN RC;
END;
```

As you can see at comment 1, you pass a parameter of size 4000. At comment 2, a buffer of size 32000 is declared. At comment 3, the buffer is filled with data (the serialized UDT). At comment 4, the buffer of size 32000 is written to the parameter of size 4000. If the data exceeds 4000 bytes, the exception occurs.

P. S.: In the method `ResetGetTypeCommand` you also set the parameter to size 4000.

---

### Workaround

I was able to work around the issue by patching the methods `CreateGetUDTTypeCmd` and `ResetGetTypeCommand` during runtime using [Harmony](https://github.com/pardeike/Harmony). I used postfix patches to overwrite the parameter size after every call to the methods.

**Comment by belaw** [comment]
I should have mentioned that this only occurs on some User Defined Types that exceed 4000 bytes when the DB pickler is trying to serialize them.

**Comment by alexkeh** [comment]
It sounds like this will be a DB bug. Since the problem doesn't happen with every UDT, can you provide your UDT type name and definition, which will give us a good idea how to reproduce your error and what causes it?

It's possible this bug may already be resolved in a DB 19c since you are using the base release. You may want to see if patching to the latest 19c RU will resolve the issue.

**Comment by belaw** [comment]
Hi @alexkeh, thanks for your Answer.

Sadly, I can't give you the UDT definition because it is part of a proprietary 3rd party product we use. I was able to make a sufficiently big dummy type though:

`BIG_UDT_SUB`
```sql
create or replace TYPE         BIG_UDT_SUB AS OBJECT 
(
    ATTRIBUTE_001 date,
    ATTRIBUTE_002 date,
    ATTRIBUTE_003 date,
    ATTRIBUTE_004 date,
    ATTRIBUTE_005 date,
    ATTRIBUTE_006 date,
    ATTRIBUTE_007 date,
    ATTRIBUTE_008 date,
    ATTRIBUTE_009 date,
    ATTRIBUTE_010 date,
    ATTRIBUTE_011 date,
    ATTRIBUTE_012 date,
    ATTRIBUTE_013 date,
    ATTRIBUTE_014 date,
    ATTRIBUTE_015 date,
    ATTRIBUTE_016 date,
    ATTRIBUTE_017 date,
    ATTRIBUTE_018 date,
    ATTRIBUTE_019 date,
    ATTRIBUTE_020 date,
    ATTRIBUTE_021 date,
    ATTRIBUTE_022 date,
    ATTRIBUTE_023 date,
    ATTRIBUTE_024 date,
    ATTRIBUTE_025 date,
    ATTRIBUTE_026 date,
    ATTRIBUTE_027 date,
    ATTRIBUTE_028 date,
    ATTRIBUTE_029 date,
    ATTRIBUTE_030 date,
    ATTRIBUTE_031 date,
    ATTRIBUTE_032 date,
    ATTRIBUTE_033 date,
    ATTRIBUTE_034 date,
    ATTRIBUTE_035 date,
    ATTRIBUTE_036 date,
    ATTRIBUTE_037 date,
    ATTRIBUTE_038 date,
    ATTRIBUTE_039 date,
    ATTRIBUTE_040 date,
    ATTRIBUTE_041 date,
    ATTRIBUTE_042 date,
    ATTRIBUTE_043 date,
    ATTRIBUTE_044 date,
    ATTRIBUTE_045 date,
    ATTRIBUTE_046 date,
    ATTRIBUTE_047 date,
    ATTRIBUTE_048 date,
    ATTRIBUTE_049 date,
    ATTRIBUTE_050 date,
    ATTRIBUTE_051 date,
    ATTRIBUTE_052 date,
    ATTRIBUTE_053 date,
    ATTRIBUTE_054 date,
    ATTRIBUTE_055 date,
    ATTRIBUTE_056 date,
    ATTRIBUTE_057 date,
    ATTRIBUTE_058 date,
    ATTRIBUTE_059 date,
    ATTRIBUTE_060 date,
    ATTRIBUTE_061 date,
    ATTRIBUTE_062 date,
    ATTRIBUTE_063 date,
    ATTRIBUTE_064 date,
    ATTRIBUTE_065 date,
    ATTRIBUTE_066 date,
    ATTRIBUTE_067 date,
    ATTRIBUTE_068 date,
    ATTRIBUTE_069 date,
    ATTRIBUTE_070 date,
    ATTRIBUTE_071 date,
    ATTRIBUTE_072 date,
    ATTRIBUTE_073 date,
    ATTRIBUTE_074 date,
    ATTRIBUTE_075 date,
    ATTRIBUTE_076 date,
    ATTRIBUTE_077 date,
    ATTRIBUTE_078 date,
    ATTRIBUTE_079 date,
    ATTRIBUTE_080 date,
    ATTRIBUTE_081 date,
    ATTRIBUTE_082 date,
    ATTRIBUTE_083 date,
    ATTRIBUTE_084 date,
    ATTRIBUTE_085 date,
    ATTRIBUTE_086 date,
    ATTRIBUTE_087 date,
    ATTRIBUTE_088 date,
    ATTRIBUTE_089 date,
    ATTRIBUTE_090 date,
    ATTRIBUTE_091 date,
    ATTRIBUTE_092 date,
    ATTRIBUTE_093 date,
    ATTRIBUTE_094 date,
    ATTRIBUTE_095 date,
    ATTRIBUTE_096 date,
    ATTRIBUTE_097 date,
    ATTRIBUTE_098 date,
    ATTRIBUTE_099 date,
    ATTRIBUTE_100 date,
    ATTRIBUTE_101 date,
    ATTRIBUTE_102 date,
    ATTRIBUTE_103 date,
    ATTRIBUTE_104 date,
    ATTRIBUTE_105 date,
    ATTRIBUTE_106 date,
    ATTRIBUTE_107 date,
    ATTRIBUTE_108 date,
    ATTRIBUTE_109 date,
    ATTRIBUTE_110 date,
    ATTRIBUTE_111 date,
    ATTRIBUTE_112 date,
    ATTRIBUTE_113 date,
    ATTRIBUTE_114 date,
    ATTRIBUTE_115 date,
    ATTRIBUTE_116 date,
    ATTRIBUTE_117 date,
    ATTRIBUTE_118 date,
    ATTRIBUTE_119 date,
    ATTRIBUTE_120 date,
    ATTRIBUTE_121 date,
    ATTRIBUTE_122 date,
    ATTRIBUTE_123 date,
    ATTRIBUTE_124 date,
    ATTRIBUTE_125 date,
    ATTRIBUTE_126 date,
    ATTRIBUTE_127 date,
    ATTRIBUTE_128 date
)
```
`BIG_UDT`
```sql
create or replace TYPE         BIG_UDT AS OBJECT 
(
    ATTRIBUTE_01 BIG_UDT_SUB,
    ATTRIBUTE_02 BIG_UDT_SUB,
    ATTRIBUTE_03 BIG_UDT_SUB,
    ATTRIBUTE_04 BIG_UDT_SUB,
    ATTRIBUTE_05 BIG_UDT_SUB,
    ATTRIBUTE_06 BIG_UDT_SUB,
    ATTRIBUTE_07 BIG_UDT_SUB,
    ATTRIBUTE_08 BIG_UDT_SUB,
    ATTRIBUTE_09 BIG_UDT_SUB,
    ATTRIBUTE_10 BIG_UDT_SUB,
    ATTRIBUTE_11 BIG_UDT_SUB,
    ATTRIBUTE_12 BIG_UDT_SUB,
    ATTRIBUTE_13 BIG_UDT_SUB,
    ATTRIBUTE_14 BIG_UDT_SUB,
    ATTRIBUTE_15 BIG_UDT_SUB
)
```

And if you need it, here's an SQL statement that can reproduce the error:
```sql
declare
    result          INT;
    FULLTYPENAME    VARCHAR2(256);
    TYPOID          RAW(16);
    VERSION         BINARY_INTEGER;
    TDS             LONG RAW(4000); -- Setting the size of this to 32000 fixes the issue
    INSTANTIABLE    VARCHAR2(256);
    SUPERTYPE_OWNER VARCHAR2(256);
    SUPERTYPE_NAME  VARCHAR2(256);
    ATTR_RC         DBMS_PICKLER.TYPE_ATTR_REF_CURSOR;
    SUBTYPE_RC      DBMS_PICKLER.TYPE_SUBTYPE_REF_CURSOR;
begin
    FULLTYPENAME := '<SCHEMA>.BIG_UDT'; -- Replace schema
    result := sys.dbms_pickler.get_type_shape(
        FULLTYPENAME,
        TYPOID,
        VERSION,
        TDS,
        INSTANTIABLE,
        SUPERTYPE_OWNER,
        SUPERTYPE_NAME,
        ATTR_RC,
        SUBTYPE_RC
    );
    DBMS_OUTPUT.PUT_LINE('result = ' || result);
    DBMS_OUTPUT.PUT_LINE('TYPOID = ' || TYPOID);
    DBMS_OUTPUT.PUT_LINE('VERSION = ' || VERSION);
    DBMS_OUTPUT.PUT_LINE('TDS Size = ' || LENGTH(TDS));
    DBMS_OUTPUT.PUT_LINE('TDS = ' || TDS);
    DBMS_OUTPUT.PUT_LINE('INSTANTIABLE = ' || INSTANTIABLE);
    DBMS_OUTPUT.PUT_LINE('SUPERTYPE_OWNER = ' || SUPERTYPE_OWNER);
    DBMS_OUTPUT.PUT_LINE('SUPERTYPE_NAME = ' || SUPERTYPE_NAME);
end;
```

**Comment by alexkeh** [comment]
Thanks! The dummy sample will be fine for us to validate a fix when we have a candidate.

**Comment by alexkeh** [comment]
The bug will be fixed in the ODP.NET 21.6 release.

**Comment by alexkeh** [comment]
The fix for this bug (33681872) is now available on NuGet Gallery.
